fix: safely initiate discord

webbertakken opened the pull request:
#### Context 

Discord initiates when reporting in tasks from CI.   
If the endpoint that was called was reportPublication 
Then discord failure would prevent image from being marked as completed 
In turn, any subsequent calls to reportPublication would fail too, because it'd already exist

Even though discord has a retry mechanism, out API key got jettisoned by Discord corp because it was hitting too many connects in an hour (and it invalidated the key automatically). This usually only happens if we develop new things in versioning-backend.

The key was replaced a while back but this probably caused a record to be updated (in `reportPublication`) after which it failed, rescheduled and tried to update again, after which the record existed.

I didn't have time to actually deep dive the logs to verify that this is in fact exactly what happened.

#### Changes

This change in combination with doing a full new run should in theory solve the problem.

- Fixes: https://github.com/game-ci/docker/pull/273
- Fixes: https://github.com/game-ci/versioning-backend/pull/73
- Fixes: https://github.com/game-ci/versioning-backend/pull/74
- Fixes: https://github.com/game-ci/docker/issues/271

#### Checklist

- [x] Read the contribution [guide](../CONTRIBUTING.md) and accept the [code](../CODE_OF_CONDUCT.md) of conduct
- [ ] Readme (updated or not needed)
- [ ] Tests (added, updated or not needed)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Safer Discord integration across build, publish, retry, cron and trigger flows to reduce initialization failures and improve resilience; more robust error handling and logging.
* **Chores**
  * Tooling update: Yarn/Volta configuration updated to a newer version.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

webbertakken committed:
fix: safely initiate discord

Discord initiates when reporting in tasks from CI.
If the endpoint that was called was reportPublication
Then discord failure would prevent image from being marked as completed
Any subsequent calls would fail too, because the job already exists

webbertakken committed:
fix: safely disconnect discord / update yarn

webbertakken committed:
fix: renew yarn.lock

webbertakken committed:
chore: update node lts

webbertakken committed:
fix: oddity with yarn

webbertakken committed:
fix: yarn immutibility

webbertakken committed:
chore: remove unnecessary yarnrc file

webbertakken committed:
chore: upgrade oxlint, fix yarn lock

webbertakken committed:
chore: lock unity-changeset to put upgrading out of scope

webbertakken committed:
chore: remove version test

webbertakken committed:
fix: derp

webbertakken commented on functions/src/service/discord.ts:
```diff
@@ -5,6 +5,14 @@ import { settings } from '../config/settings';
 let instance: ErisClient | null = null;
 let instanceCount = 0;
 export class Discord {
+  public static async initSafely(token: string) {
+    try {
+      await this.init(token);
+    } catch (err) {
+      logger.error('Failed to initialize Discord client safely:', err);
+    }
+  }
```
Nah that's alright

webbertakken committed:
chore: restore immutable

webbertakken committed:
chore: clenaup

GabLeRoux commented:
<del>The test deploy action failed, but other than that,</del> this looks good to me.

I challenged this change a bit on my side, but it's most likely OK to ship this as is. I'm also OK with the fire-and-forget part about the Discord bot (CodeRabbit's comment).

Here's a nitpick that we could probably do since the code change is minimal, but it eventually gets cleaned up anyway.

<details><summary>Review & Challenge from Claude Opus 4.5</summary>
<p>

# Discord Init Review

## Concern

When `init()` fails after creating the ErisClient but before incrementing `instanceCount`, the state becomes inconsistent:

- `instance` holds a broken/partial client
- `instanceCount` remains 0

Subsequent calls see `instance` as truthy, increment the count, and return early—reusing a broken client that will never send messages.

## Sequence Diagram

```mermaid
sequenceDiagram
    participant A as Request A
    participant B as Request B
    participant D as Discord Service
    participant E as ErisClient

    Note over D: instance = null<br/>instanceCount = 0

    A->>D: initSafely(token)
    D->>D: init(token)
    D->>E: new ErisClient(token)
    Note over D: instance = ErisClient
    D->>E: connect()
    E-->>D: timeout / error
    Note over D: throws "Bot never became ready"
    D-->>A: error caught, logged
    Note over D: instance = broken client ❌<br/>instanceCount = 0

    B->>D: initSafely(token)
    D->>D: init(token)
    Note over D: instance is truthy → early return
    D->>D: instanceCount += 1
    Note over D: instanceCount = 1
    D-->>B: returns (no error)

    B->>D: sendAlert("message")
    D->>E: createMessage()
    Note over E: broken client, message lost silently
```

## Real Impact

In practice, **low risk**:

1. **Cloud Functions are ephemeral** — each invocation typically gets a fresh container. The zombie state only persists within a single warm instance.

2. **Failure is rare** — this only happens if Discord init fails (rate limiting, network issues). Under normal operation, the path is never hit.

3. **Already "fire and forget"** — with `initSafely`, you've accepted that Discord failures won't block business logic. A zombie instance just means continued silent failure until the container recycles.

4. **Self-healing** — when the container eventually cold-starts, state resets.

## Verdict

**Safe to ship as-is.** The edge case is unlikely and self-corrects on container recycle.

However, the fix is trivial and makes the code more correct. Worth adding if you're already touching this code:

```typescript
public static async init(token: string) {
  if (instance) {
    instanceCount += 1;
    return;
  }

  instance = new ErisClient(token);
  try {
    await instance.connect();
    // ... wait for ready ...
    instanceCount += 1;
  } catch (err) {
    instance = null;  // reset on failure
    throw err;
  }
}
```

</p>
</details>

webbertakken commented:
I see there are no messages coming to discord at all.

Looking at the logs from the cron

<img width="1679" height="501" alt="image" src="https://github.com/user-attachments/assets/97f82d75-e6ac-4f46-bfb2-24e8d679a962" />

webbertakken commented:
I had updated the secrets incorrectly, following our own docs that was relevant for legacy firebase versions.

<img width="841" height="58" alt="image" src="https://github.com/user-attachments/assets/283290c0-d94e-483c-93d7-1ac52b480e39" />

I've updated it again and logs look good now

<img width="665" height="481" alt="image" src="https://github.com/user-attachments/assets/90572cdb-c789-4f0e-bd40-f275d4641d07" />

Will update docs in https://github.com/game-ci/versioning-backend/pull/76

webbertakken commented:
Builds are reporting in on [our Discord server](http://game.ci/discord)

<img width="1090" height="183" alt="image" src="https://github.com/user-attachments/assets/c9108347-7fae-4257-b40a-618c5b22dc57" />

Completed versions can be seen on the [versions page](https://game.ci/docs/docker/versions/) of our docs website

<img width="624" height="225" alt="image" src="https://github.com/user-attachments/assets/8d55bd3f-4726-48ad-a321-4237b3db652a" />


They can be seen on the [actions tab](https://github.com/game-ci/docker/actions) of our docker repo:

<img width="2111" height="693" alt="image" src="https://github.com/user-attachments/assets/ba5e3a39-ae33-44ee-895d-688b8885f506" />
